### DISCUSSION TITLE
Different stored procedures depending on database schema

### DISCUSSION THREAD

**Discussion by MapInetum** [Mimetis/Dotmim.Sync#1286] [Q&A] [answered]
I have encountered this problem using version 1.2.0 of the DMS framework and the provider 'SqlSyncChangeTrackingProvider'.

Provisioning a table located in the default schema of the database (in this case 'dbo') using a 'CustomWhere' to filter the synchronization by a particular field, it is observed that the following stored procedures are created, including those necessary to select the changes by the configured filter.

![image](https://github.com/user-attachments/assets/ab7ea4aa-86fe-44a2-ba3f-f5cf3b45be53)
![image](https://github.com/user-attachments/assets/71417ce3-fec5-48ef-a78e-983ca37abebe)
![image](https://github.com/user-attachments/assets/66ecb541-f9a1-41df-91fb-978b5ccaf903)
![image](https://github.com/user-attachments/assets/93fc4be6-6d02-49e7-aaa2-ab196cf46fee)

On the other hand, if the table to be synchronized is in another schema of the database (such as the 'js' schema), when provisioning the table in the same way and with the same filter, the necessary stored procedures are not created to filter the changes by the configured parameter.

![image](https://github.com/user-attachments/assets/6eb9c0be-25d2-49ef-865b-811f41cc19c2)
![image](https://github.com/user-attachments/assets/309d2188-2128-45ba-8254-ecbde0a5420f)
![image](https://github.com/user-attachments/assets/5057562e-4a0f-45e7-8a66-ac9349282f87)
![image](https://github.com/user-attachments/assets/0bae0bc7-0172-46b4-b981-6b84dd6d7dea)

Any idea why this is happening?

I need to use the 'js' schema to provision all the procedures to be able to filter by the configured parameter.

**Comment by MapInetum** [comment]
Can you help me please?

**Reply by Mimetis** [reply]
I don't understand your question.
Best thing to do is to create a sample that you can share
Create a github repository with a very simple console application that describe in  code what you want to do (and where it's failing), a sql schema I can use to create the server db, and a readme file to explains the process

**Reply by MapInetum** [reply]
Hi.
I have created a repository with a console project where I reproduce what I explain above.

https://github.com/maprado-inetum/DotmimSyncExamples

Importing first in SQL Server the two databases (ServerDatabase and ClientDatabase) and running the offered console, you can see how in the generated stored procedures are missing those that filter by the condition written in the 'CustomWhere'.

**Reply by Mimetis** [reply]
Hey @maprado-inetum
Thanks a lot to provide a sample, much appreciated !

Here is the sql setup I used (based on your backups):

``` sql
USE [master]
GO 
-- Server database
if (exists (select * from sys.databases where name = 'ServerDatabase'))
Begin
	ALTER DATABASE [ServerDatabase] SET  SINGLE_USER WITH ROLLBACK IMMEDIATE;
	DROP DATABASE [ServerDatabase]
End
Create database [ServerDatabase]
Go
-- Client database. No need to create the schema, Dotmim.Sync will do
if (exists (select * from sys.databases where name = 'ClientDatabase'))
Begin
	ALTER DATABASE [ClientDatabase] SET  SINGLE_USER WITH ROLLBACK IMMEDIATE;
	DROP DATABASE [ClientDatabase]
End
Create database [ClientDatabase]
GO
USE [ServerDatabase]
GO
CREATE SCHEMA [js]
GO
CREATE TABLE [js].[OnDemandJobs](
	[Id] [uniqueidentifier] NOT NULL PRIMARY KEY  DEFAULT (newsequentialid()),
	[AssetAlias] [nvarchar](max) NULL,
	[Name] [nvarchar](max) NULL,
	[Server] [nvarchar](max) NULL,
	[User] [nvarchar](max) NULL,
	[Parameters] [nvarchar](max) NULL,
	[Executed] [bit] NOT NULL,
	[ExecutionDurationInSeconds] [int] NULL,
	[ExecutionTimeUtc] [datetime2](7) NULL,
	[ExecutionMessage] [nvarchar](max) NULL
)
GO

USE [ClientDatabase]
GO
CREATE SCHEMA [js]
GO
CREATE TABLE [js].[OnDemandJobs](
	[Id] [uniqueidentifier] NOT NULL PRIMARY KEY DEFAULT (newsequentialid()),
	[AssetAlias] [nvarchar](max) NULL,
	[Name] [nvarchar](max) NULL,
	[Server] [nvarchar](max) NULL,
	[User] [nvarchar](max) NULL,
	[Parameters] [nvarchar](max) NULL,
	[Executed] [bit] NOT NULL,
	[ExecutionDurationInSeconds] [int] NULL,
	[ExecutionTimeUtc] [datetime2](7) NULL,
	[ExecutionMessage] [nvarchar](max) NULL
) 
GO
ALTER DATABASE [ServerDatabase] SET ALLOW_SNAPSHOT_ISOLATION ON  
ALTER DATABASE [ServerDatabase] SET READ_COMMITTED_SNAPSHOT ON  
ALTER DATABASE [ServerDatabase] SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 14 DAYS, AUTO_CLEANUP = ON)

ALTER DATABASE [ClientDatabase] SET ALLOW_SNAPSHOT_ISOLATION ON  
ALTER DATABASE [ClientDatabase] SET READ_COMMITTED_SNAPSHOT ON  
ALTER DATABASE [ClientDatabase] SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 14 DAYS, AUTO_CLEANUP = ON)
```
I didn't changed the code, as it's running correctly.

And, indeed, the stored procedures are not created, because the `SetupFilter `object should take, as parameters, the **table name** and the **schema name**:

``` cs
SetupFilter onDemandJobsFilter = new("OnDemandJobs", "js");
```

That way, the stored procedures are correctly created:

![image](https://github.com/user-attachments/assets/55bd3640-3cf2-4657-8aa8-5fc15c9cb9e6)


``` sql
USE [ServerDatabase]
GO
/****** Object:  StoredProcedure [js].[OnDemandJobs_OnDemandJobsSync_InstallationAliaschanges]    Script Date: 18/12/2024 09:28:15 ******/
SET ANSI_NULLS ON
GO
SET QUOTED_IDENTIFIER ON
GO
ALTER PROCEDURE [js].[OnDemandJobs_OnDemandJobsSync_InstallationAliaschanges]
	@sync_min_timestamp bigint,
	@sync_scope_id uniqueidentifier,
	@InstallationAlias nvarchar (MAX)
AS
BEGIN
;WITH 
  [OnDemandJobs_tracking] AS (
	SELECT [CT].[Id], 
	CAST([CT].[SYS_CHANGE_CONTEXT] AS uniqueidentifier) AS [sync_update_scope_id],
	[CT].[SYS_CHANGE_VERSION] AS [sync_timestamp],
	CASE WHEN [CT].[SYS_CHANGE_OPERATION] = 'D' THEN 1 ELSE 0 END AS [sync_row_is_tombstone],
	[CT].[SYS_CHANGE_COLUMNS] AS [sync_change_columns]
	FROM CHANGETABLE(CHANGES [js].[OnDemandJobs], @sync_min_timestamp) AS [CT]
	)
SELECT DISTINCT
	[side].[Id], 
	[base].[AssetAlias], 
	[base].[Name], 
	[base].[Server], 
	[base].[User], 
	[base].[Parameters], 
	[base].[Executed], 
	[base].[ExecutionDurationInSeconds], 
	[base].[ExecutionTimeUtc], 
	[base].[ExecutionMessage], 
	[side].[sync_row_is_tombstone],
	[side].[sync_update_scope_id]
FROM [js].[OnDemandJobs] [base]
RIGHT JOIN [OnDemandJobs_tracking] [side]ON [base].[Id] = [side].[Id]
WHERE (
(
  [base].[AssetAlias] = @InstallationAlias OR [side].[sync_row_is_tombstone] = 1
)
AND 
	[side].[sync_timestamp] > @sync_min_timestamp
	AND ([side].[sync_update_scope_id] <> @sync_scope_id OR [side].[sync_update_scope_id] IS NULL)
)

END
```

Let me know if it's resolving your issue.

**Reply by MapInetum** [reply]
Thank you for the contribution and the quick response, but I am very sorry to say that the procedures are not created for me:

- [js].[OnDemandJobs_OnDemandJobsSync_InstallationAliasinitialize]
- [js].[OnDemandJobs_OnDemandJobsSync_InstallationAliaschanges]

Could the problem lie in the SQL Server version? I use 14.0.3485.1

I have uploaded to the test repository the change to add the schema in the 'setupFilter' for you to see.
I have also replaced the .bak with the script you have shared.
Thanks

**Reply by Mimetis** [reply, marked as answer]
that's incorrect:

``` cs
 string tableName = $"js.OnDemandJobs";

// Filters
SetupFilter onDemandJobsFilter = new(tableName, "js");
```

Should be :

``` cs
 string tableName = $"js.OnDemandJobs";

// Filters
SetupFilter onDemandJobsFilter = new("OnDemandJobs", "js");
```

Or even


``` cs
string tableName = $"OnDemandJobs";
string schemaName = $"js";

// Filters
SetupFilter onDemandJobsFilter = new(tableName, schemaName);
```

**Reply by MapInetum** [reply]
It works! I have uploaded to the repository the change I made to make it work.
Thanks a lot!
